drivers/pwm/pwm_ifx_cat1.c

```diff
@@ -0,0 +1,177 @@
+#include <cy_sysclk.h>
+#include <cyhal_hw_resources.h>
+#include <cyhal_hw_types.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pwm_ifx_cat1, CONFIG_PWM_LOG_LEVEL);
+
+#define PWM_REG_BASE TCPWM0
+
+struct ifx_cat1_pwm_data {
+	uint32_t pwm_num;
+};
+
+struct ifx_cat1_pwm_config {
+	TCPWM_GRP_CNT_Type *reg_addr;
+	const struct pinctrl_dev_config *pcfg;
+	bool resolution_32_bits;
+	cy_en_divider_types_t divider_type;
+	uint32_t divider_sel;
+	uint32_t divider_val;
+};
+
+static int ifx_cat1_pwm_init(const struct device *dev)
+{
+	struct ifx_cat1_pwm_data *data = dev->data;
+	const struct ifx_cat1_pwm_config *config = dev->config;
+	cy_en_tcpwm_status_t status;
+	int ret;
+	uint32_t addr_offset = (uint32_t)config->reg_addr - TCPWM0_BASE;
+	uint32_t clk_connection;
+
+	const cy_stc_tcpwm_pwm_config_t pwm_config = {
+		.pwmMode = CY_TCPWM_PWM_MODE_PWM,
+		.clockPrescaler = CY_TCPWM_PWM_PRESCALER_DIVBY_1,
+		.pwmAlignment = CY_TCPWM_PWM_LEFT_ALIGN,
+		.runMode = CY_TCPWM_PWM_CONTINUOUS,
+		.countInputMode = CY_TCPWM_INPUT_LEVEL,
+		.countInput = CY_TCPWM_INPUT_1,
+	};
+
+	/* Configure PWM clock */
+	Cy_SysClk_PeriphDisableDivider(config->divider_type, config->divider_sel);
+	Cy_SysClk_PeriphSetDivider(config->divider_type, config->divider_sel, config->divider_val);
+	Cy_SysClk_PeriphEnableDivider(config->divider_type, config->divider_sel);
+
+	/* This is very specific to the cyw920829m2evk_02 and may need to be modified
+	 * for other boards.
+	 */
+	if (addr_offset < sizeof(TCPWM_GRP_Type)) {
+		clk_connection =
+			PCLK_TCPWM0_CLOCK_COUNTER_EN0 + (addr_offset / sizeof(TCPWM_GRP_CNT_Type));
+	} else {
+		addr_offset -= sizeof(TCPWM_GRP_Type);
+		clk_connection = PCLK_TCPWM0_CLOCK_COUNTER_EN256 +
+				 (addr_offset / sizeof(TCPWM_GRP_CNT_Type));
+	}
+	Cy_SysClk_PeriphAssignDivider(clk_connection, config->divider_type, config->divider_sel);
+
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Configure the TCPWM to be a PWM */
+	data->pwm_num += addr_offset / sizeof(TCPWM_GRP_CNT_Type);
+	status = Cy_TCPWM_PWM_Init(PWM_REG_BASE, data->pwm_num, &pwm_config);
+	if (status != CY_TCPWM_SUCCESS) {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int ifx_cat1_pwm_set_cycles(const struct device *dev, uint32_t channel,
+				   uint32_t period_cycles, uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	struct ifx_cat1_pwm_data *data = dev->data;
+	const struct ifx_cat1_pwm_config *config = dev->config;
+
+	if (!config->resolution_32_bits &&
+	    ((period_cycles > UINT16_MAX) || (pulse_cycles > UINT16_MAX))) {
+		/* 16-bit resolution */
+		if (period_cycles > UINT16_MAX) {
+			LOG_ERR("Period cycles more than 16-bits (%u)", period_cycles);
+		}
+		if (pulse_cycles > UINT16_MAX) {
+			LOG_ERR("Pulse cycles more than 16-bits (%u)", pulse_cycles);
+		}
+		return -EINVAL;
+	}
+
+	if ((period_cycles == 0) || (pulse_cycles == 0)) {
+		Cy_TCPWM_PWM_Disable(PWM_REG_BASE, data->pwm_num);
```

Comment:
The PWM is disabled, but shouldn't its output be set to the inactive state in accordance with the passed polarity flag?